influxql/functions.go

```diff
@@ -309,3 +316,322 @@ func (r *IntegerMovingAverageReducer) Emit() []FloatPoint {
 		},
 	}
 }
+
+// FloatHoltWintersReducer forecasts a series into the future.
+// This is done using the Holt-Winters damped method.
+//    1. Using the series the initial values are calculated using a SSE.
+//    2. The series is forecasted into the future using the iterative relations.
+type FloatHoltWintersReducer struct {
+	// Smoothing parameters
+	alpha,
+	beta,
+	gamma float64
+
+	// Dampening parameter
+	phi float64
+
+	// Season period
+	m        int
+	seasonal bool
+
+	// Horizon
+	h int
+
+	// Interval between points
+	interval int64
+
+	// Whether to include all data or only future values
+	includeFitData bool
+
+	// NelderMead optimizer
+	optim *neldermead.Optimizer
+	// Small difference bound for the optimizer
+	epsilon float64
+
+	y      []float64
+	points []FloatPoint
+}
+
+const (
+	defaultAlpha   = 0.5
+	defaultBeta    = 0.5
+	defaultGamma   = 0.5
+	defaultPhi     = 0.5
+	defaultEpsilon = 1.0e-4
+)
+
+// NewFloatHoltWintersReducer creates a new FloatHoltWintersReducer.
+func NewFloatHoltWintersReducer(h, m int, includeFitData bool, interval time.Duration) *FloatHoltWintersReducer {
```

Comment:
I don't see any check that the horizon is positive. If it's not checked then a bunch of places can panic where you're allocating slices based on the size of the horizon.